test/modules/TestResultCoordinator/Reports/EdgeHubRestartTest/EdgeHubRestartDirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,304 @@
+
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Generating report: {nameof(EdgeHubRestartDirectMethodReport)} for [{this.Metadata.SenderSource}] and [{this.Metadata.ReceiverSource}]");
+
+            bool isPassing = true;
+            long previousSeqNum = 0;
+            long passedDirectMethodCount = 0;
+
+            // Value: (source, numOfResults)
+            Dictionary<string, ulong> resultCount = new Dictionary<string, ulong>()
+            {
+                { nameof(this.SenderTestResults), 0ul },
+                { nameof(this.ReceiverTestResults), 0ul }
+            };
+
+            // Value: (completedStatusCode, DirectMethodCompletedTime - EdgeHubRestartedTime)
+            Dictionary<HttpStatusCode, List<TimeSpan>> completedStatusHistogram = new Dictionary<HttpStatusCode, List<TimeSpan>>();
+
+            bool hasSenderResult = await this.SenderTestResults.MoveNextAsync();
+            bool hasReceiverResult = await this.ReceiverTestResults.MoveNextAsync();
+
+            while (hasSenderResult && hasReceiverResult)
+            {
+                this.ValidateResult(
+                    this.SenderTestResults.Current,
+                    this.Metadata.SenderSource,
+                    TestOperationResultType.EdgeHubRestartDirectMethod.ToString());
+
+                this.ValidateResult(
+                    this.ReceiverTestResults.Current,
+                    this.Metadata.ReceiverSource,
+                    TestOperationResultType.DirectMethod.ToString());
+
+                // Both sender & receiver have their dm results
+                resultCount[nameof(this.SenderTestResults)]++;
+                resultCount[nameof(this.ReceiverTestResults)]++;
+
+                // Adjust seqeunce number from both source to be equal before doing any comparison
+                EdgeHubRestartDirectMethodResult senderResult = JsonConvert.DeserializeObject<EdgeHubRestartDirectMethodResult>(this.SenderTestResults.Current.Result);
+                DirectMethodTestResult receiverResult = JsonConvert.DeserializeObject<DirectMethodTestResult>(this.ReceiverTestResults.Current.Result);
+
+                long senderSeqNum = this.ConvertStringToLong(senderResult.SequenceNumber);
+                long receiverSeqNum = this.ConvertStringToLong(receiverResult.SequenceNumber);
+
+                if (receiverSeqNum > senderSeqNum)
+                {
+                    // Increment sender result to have the same seq as the receiver
+                    await this.IncrementSequenceNumberAsync(
+                        this.SenderTestResults,
+                        nameof(this.SenderTestResults),
+                        receiverSeqNum,
+                        resultCount,
+                        completedStatusHistogram);
+
+                    // Fail the test
+                    isPassing = false;
+                }
+
+                if (receiverSeqNum < senderSeqNum)
```

Comment:
can you tell me any possibilities for this condition become true?

Reply:
In normal scenario, it shouldn't happen. I can think if somehow edgeHub retries a directMethod (which shouldn't happen).

Reply:
General comment is to simplify report generation logic; for example, the flow should try focusing on happy path and fail all other cases generally.  If code get complex due to distinguish some edge cases failed results, I would prefer not distinguish them; fail it and report all failed data and investigate later.

Reply:
+1 on that. Please keep tests as simple as possible.